File: src/test_programs/F03/test_geomgrad.f90
program test_geomgrad
    use iso_c_binding, only: c_char
    use iso_fortran_env, only: error_unit
    use ommp_interface

    implicit none

    integer :: argc, i, retcode=0, io_file=101, mm_atoms
    character(kind=c_char, len=120), dimension(3) :: argv
    real(ommp_real), allocatable :: grad_ana(:,:)
    type(ommp_system), pointer :: my_system

    argc = command_argument_count()
    if(argc /= 2) then
        write(*, *) "Syntax expected"
        write(*, *) "    $ test_init.exe <INPUT FILE> <OUTPUT FILE>"
        stop 1
    end if

    do i=1, argc
        call get_command_argument(i, argv(i))
    end do

    call ommp_init_mmp(my_system, argv(1))
    mm_atoms = my_system%top%mm_atoms

    allocate(grad_ana(3, my_system%top%mm_atoms))
    open(io_file, file=argv(2))

    grad_ana = 0.0
    call ommp_fixedelec_geomgrad(my_system, grad_ana)

    write(*, *) "Grad EM"
    do i=1, mm_atoms
        write(io_file, "(3f12.8)")grad_ana(:,i)
    end do
    
    grad_ana = 0.0
    call ommp_polelec_geomgrad(my_system, grad_ana)

    write(*, *) "Grad EP"
    do i=1, mm_atoms
        write(io_file, "(3f12.8)")grad_ana(:,i)
    end do
    
    deallocate(grad_ana)
    close(io_file)

    call ommp_terminate(my_system)
    
    if(retcode > 0) then
        stop 1
    else
        stop 0
    end if
end program

File: src/test_programs/C/test_geomgrad.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include "openmmpol.h"

int main(int argc, char **argv){
    if(argc != 3){
        printf("Syntax expected\n");
        printf("    $ test_geomgrad.exe <INPUT FILE> <OUTPUT FILE>\n");
        return 0;
    }
    
    int pol_atoms, mm_atoms, retcode=0;
    double **grad_ana, *_grad_ana;

    OMMP_SYSTEM_PRT my_system = ommp_init_mmp(argv[1]);
    ommp_set_verbose(OMMP_VERBOSE_NONE);
    
    pol_atoms = ommp_get_pol_atoms(my_system);
    mm_atoms = ommp_get_mm_atoms(my_system);
    
    _grad_ana = (double *) malloc(sizeof(double) * 3 * mm_atoms);
    grad_ana = (double **) malloc(sizeof(double *) * mm_atoms);
    for(int i = 0; i < mm_atoms; i++){
        grad_ana[i] = &(_grad_ana[i*3]);
        for(int j=0; j < 3; j++)
            grad_ana[i][j] = 0.0;
    }
   
    FILE *fp = fopen(argv[2], "w+");
    
    ommp_fixedelec_geomgrad(my_system, _grad_ana);
    
    fprintf(fp, "Grad EM\n");
    for(int i = 0; i < mm_atoms; i++){
        for(int j=0; j < 3; j++)
            fprintf(fp, "%+12.8g ", grad_ana[i][j]);
        fprintf(fp, "\n");
    }
    fprintf(fp, "\n");

    for(int i = 0; i < mm_atoms; i++)
        for(int j=0; j < 3; j++)
            grad_ana[i][j] = 0.0;
    
    ommp_polelec_geomgrad(my_system, _grad_ana);
   
    fprintf(fp, "Grad EP\n");
    for(int i = 0; i < mm_atoms; i++){
        for(int j=0; j < 3; j++)
            fprintf(fp, "%+12.8g ", grad_ana[i][j]);
        fprintf(fp, "\n");
    }
    fprintf(fp, "\n");

    fclose(fp);
    ommp_terminate(my_system);
    
    return retcode;
}

